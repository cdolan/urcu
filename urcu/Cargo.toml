[package]
name = "urcu2"
license = "GPL-3.0-or-later"
version = "0.1.0"
edition = "2021"

[lib]
name = "urcu"
doctest = false

[package.metadata.docs.rs]
features = ["static"]

[dependencies]
anyhow = "1"
container_of = "0.5"
guardian = "1"
libc = "0.2"
log = "0.4"
paste = "1"
static_assertions = "1"
urcu2-bp-sys = { path = "../urcu-bp-sys", optional = true }
urcu2-mb-sys = { path = "../urcu-mb-sys", optional = true }
urcu2-memb-sys = { path = "../urcu-memb-sys", optional = true }
urcu2-qsbr-sys = { path = "../urcu-qsbr-sys", optional = true }
urcu2-sys = { path = "../urcu-sys" }

[dev-dependencies]
ctrlc = "3"
env_logger = "0.11"
rand = "0.8"

[features]
default = [ "flavor-bp", "flavor-mb", "flavor-memb", "flavor-qsbr" ]
flavor-bp = ["dep:urcu2-bp-sys"]
flavor-mb = ["dep:urcu2-mb-sys"]
flavor-memb = ["dep:urcu2-memb-sys"]
flavor-qsbr = ["dep:urcu2-qsbr-sys"]
static = [
    "urcu2-bp-sys/static",
    "urcu2-mb-sys/static",
    "urcu2-memb-sys/static",
    "urcu2-qsbr-sys/static",
    "urcu2-sys/static",
]

