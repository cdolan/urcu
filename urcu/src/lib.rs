#![feature(ptr_as_ref_unchecked)]
#![doc = include_str!("../../README.md")]

mod rcu;
mod utility;

pub mod boxed;
pub mod hashmap;
pub mod list;
pub mod queue;
pub mod shared;
pub mod stack;

pub use crate::boxed::container::RcuBox;
pub use crate::hashmap::container::RcuHashMap;
pub use crate::list::container::RcuList;
pub use crate::queue::container::RcuQueue;
pub use crate::rcu::callback::*;
pub use crate::rcu::cleanup::*;
pub use crate::rcu::flavor::DefaultFlavor;
pub use crate::rcu::flavor::RcuFlavor;
pub use crate::rcu::guard::RcuGuard;
pub use crate::rcu::poller::RcuPoller;
pub use crate::rcu::reference::*;
pub use crate::rcu::*;
pub use crate::stack::container::RcuStack;

pub mod prelude {
    pub use crate::{RcuFlavor, RcuGuard, RcuPoller, RcuRef};

    pub use crate::{RcuContext, RcuDeferContext, RcuReadContext};

    pub use crate::{RcuBox, RcuHashMap, RcuList, RcuQueue, RcuStack};

    pub use crate::{DefaultContext, DefaultFlavor};
}
