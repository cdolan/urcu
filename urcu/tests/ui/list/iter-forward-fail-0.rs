use urcu::context::RcuContextMemb;
use urcu::{RcuContext, RcuReadContext};
use urcu::RcuList;

fn main() {
    let context = RcuContextMemb::rcu_register().unwrap();
    
    let list = RcuList::<u32, RcuContextMemb>::new();
    let guard = context.rcu_read_lock();
    let mut iter = list.iter_forward(&guard);
    drop(list);
    log::info!("{:?}", iter.next());
    drop(guard);
}
