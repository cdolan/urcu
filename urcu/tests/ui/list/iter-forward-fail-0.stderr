error[E0505]: cannot move out of `list` because it is borrowed
  --> tests/ui/list/iter-forward-fail-0.rs:9:10
   |
6  |     let list = RcuList::<u32>::new();
   |         ---- binding `list` declared here
7  |     let guard = context.rcu_read_lock();
8  |     let mut iter = list.iter_forward(&guard);
   |                    ---- borrow of `list` occurs here
9  |     drop(list);
   |          ^^^^ move out of `list` occurs here
10 |     log::info!("{:?}", iter.next());
   |                        ---- borrow later used here
